Display multiple categories

danchamp opened the pull request:
This PR:

- Closes #13 #6 #16 #20 #24 #23 
- Displays multiple news categories on a news article
- Implements [entity_reference_facet_link](https://www.drupal.org/project/entity_reference_facet_link) to link categories to facet search
- Adds subscriber to localgov_core PageHeaderDisplayEvent to suppress page header block
- Improves article and newsroom templates to follow Drupal standards more closely and use BEM class naming
- Suppresses the 'Show more' button on initial load if not required

Thanks @markconroy

danchamp committed:
Show multiple categories, improve semantics

danchamp committed:
CSS for news article categories

danchamp commented:
@finnlewis @andybroomfield This was the issue I mentioned in a recent stand-up - how to amend these categories to use the facet path instead of the taxonomy term path.

I'm aware of the facet URL generator service (facets.utility.url_generator) but can't figure out where to use it to alter the category links, or if there's an easier way.

Any pointers much appreciated.

danchamp committed:
Link news categories to facet filters if they exist

danchamp committed:
Check field and facet exist before processing categories

danchamp committed:
Coding standards fix

j4-m commented:
Hey @danchamp not really sure about the context of this PR but I had it on my list to review this from our tech meeting last week - my initial impressions are 1) that it seems strange that Facets doesn't handle this behaviour by default and 2) was that it seemed a bit heavy for a preprocess and maybe we could make use of the Facets API instead if something custom is needed. 

Could you give me a little more insight into the problem we're trying to solve? In the meantime, I'll get the project set up locally.

danchamp commented:
Thanks @j4-m.

News has an entity reference field `localgov_news_categories` that uses the Localgov Topics taxonomy.

For news articles we want referenced categories listed and linked to the facet URL, not the taxonomy URL.

You can see this on the London Councils dev site (where this branch is being used) and Crime and public protection links to the facet:

https://dev.londoncouncils.gov.uk/news/2021/bacon

I'm not very experienced with Drupal 8 at this level, and struggled to find out how to do this. Everything pointed to the FacetsUrlGenerator service (facets.utility.url_generator), but I'm not sure how best to use it - using node preprocess works, but I have no idea about what alternatives there are or implications of doing it this way.

Any help / pointers much appreciated.

j4-m commented:
Thanks for clarifying @danchamp. Have you come across [entity_reference_facet_link](https://www.drupal.org/project/entity_reference_facet_link)? I just stumbled across this while looking for something else and seems like it might be useful...

markconroy commented:
Just for reference, I am approaching the same issue but doing it in the Twig template instead of a preprocess function, here https://github.com/localgovdrupal/localgov_news/pull/22/files#diff-48c85205507538984bcd3a2447d2538e460aa4c78e8db43d1a74c4fa945d82deR87

I'm not sure which is the better approach. I like to keep things in the Twig template so everything I need is in one place and not spread out across multiple files.

danchamp commented:
Thanks @j4-m, I had seen that but wasn't sure about using a module for what seemed to be such a minor requirement, or one that isn't covered by the security advisory policy. 

@markconroy I was resistant to hard-coding the facet links in case the facet wasn't active, but it's probably such an edge case that doing it in twig is a fair approach, and much lower weight.

j4-m commented:
As the module seems to cater for the exact use case my preference would be to use contrib. especially as @markconroy also has this requirement - better to consolidate the implementation which can be tested and reused.

danchamp committed:
Implement Entity Feference Facet Link module

Update news article template to follow Drupal standards and use BEM classes

danchamp committed:
Fix tests failing with missing schema error

danchamp committed:
Remove unused use statements

danchamp committed:
Remove default_config_hash from config

danchamp commented:
@j4-m @markconroy thanks for your help. This is now updated to use the [Entity Reference Facet Link](https://www.drupal.org/project/entity_reference_facet_link) module, and with the [template updates](https://github.com/localgovdrupal/localgov_news/pull/22/files#diff-48c85205507538984bcd3a2447d2538e460aa4c78e8db43d1a74c4fa945d82deR87) from Mark's fork incorporated.

Please let me know if this looks ready for review. Once this is merged I'll remove the CSS from the module, update localgov_theme to reflect the class name changes, and hopefully get the rest of the outstanding issues sorted.

danchamp commented:
> With that module can facet point to different paths depending on where it is used? I'm not even sure what I am asking but I'm just wondering if there's a case where we would want the facet to point to /events/... in some places but /news/... in others? Maybe there's not this use case.

The module provides for that in the UI, when you select facet link as the format you can select which facet source to use.

I created a duplicate of the news search view and facets to demonstrate:

<img width="1207" alt="Screenshot 2021-04-09 at 11 29 35" src="https://user-images.githubusercontent.com/261421/114167632-1761ed00-9927-11eb-932b-caec843b8224.png">

<img width="562" alt="Screenshot 2021-04-09 at 11 29 44" src="https://user-images.githubusercontent.com/261421/114167644-1c26a100-9927-11eb-82ec-a713ec0d5290.png">

markconroy commented on localgov_news.module:
```diff
@@ -24,6 +24,14 @@ function localgov_news_theme($existing, $type, $theme, $path) {
+ */
+function localgov_news_page_attachments(array &$attachments) {
+  // Attach news styling library.
+  $attachments['#attached']['library'][] = 'localgov_news/news-styling';
```
Does this mean that the library gets attached to every page once the news module is installed? 

If so, we should amend it to only attach the new styling to pages where the content type is news-article and/or newsroom.

danchamp replied:
> Does this mean that the library gets attached to every page once the news module is installed?
> 
> If so, we should amend it to only attach the new styling to pages where the content type is news-article and/or newsroom.

My intention was to remove all styling from the news module. There's only a very basic layout for the newsroom, and the category styling in the module, so it makes little sense to have it here?

This is definitely an area where some standards for the project would help.

j4-m replied:
@danchamp if you do need to keep some styles with this module you can use `hook_ENTITY_TYPE_view_alter` with a condition on the bundle only attach to news nodes.

danchamp replied:
Thanks @j4-m. Is there any convention for when to include CSS (or what CSS to include) within a module?

I can see that most localgov_ modules have none or very little, except step-by-step and subsites both of which have interface elements.

j4-m replied:
@danchamp generally, and particularly with contrib., a module would only provide CSS if it implements it's [theme hook](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21theme.api.php/function/hook_theme/8.2.x) and this is the case us [step-by-step](https://github.com/localgovdrupal/localgov_step_by_step/blob/master/localgov_step_by_step.module#L21) and [subsites](https://github.com/localgovdrupal/localgov_subsites/blob/master/localgov_subsites.module#L14).

If your module is not implementing hook theme, i.e. not providing a new template, then ideally you would be making use of templates provided by core or contrib., via field formatters etc., and leaving the styles to the theme.

LocalGov is a bit of a different beast however and really sure what the precedent is for this sort of thing. @markconroy, is this the sort of thing you're hoping to work through in the meeting next Thursday?

Hope this helps @danchamp  Do let me know if it's not clear.

markconroy commented on templates/node--localgov-news-article--full.html.twig:
```diff
@@ -67,23 +66,41 @@
-    </div>
+  <div class="news-article__metadata">
+    <time datetime="{{ node.localgov_news_date()|date("U") }}" class="news-article__metadata-item news-article__metadata--date">
+      {{ node.localgov_news_date()|date("j F Y") }}
```
For this particular item, I'd like if we rendered the `localgov_news_date` field via the "Manage display" page with a default formatter of "j F Y". That would mean each site could override this without needed to maintain their own copy of this template. If we hardcode the date format in the template, then every theme must use the same format or else clone this template and maintain their own version of it.

markconroy commented on templates/node--localgov-news-article--full.html.twig:
```diff
@@ -67,23 +66,41 @@
+      {{ node.localgov_news_date()|date("j F Y") }}
+    </time>
+    {% if not node.localgov_news_categories.isEmpty() %}
+      <ul class="news-article__metadata-item news-article__metadata--categories" aria-label="Article categories">
```
Let's put this aria-label into a translatable string like so:

`aria-label="{{ 'Article categories'|t }}"`

This will be helpful for any councils that need multilingual, such as Scottish or Welsh ones. Even if none need to be multilingual, it's still a Drupal best practice to have that facility there.

markconroy commented on templates/node--localgov-news-article--full.html.twig:
```diff
@@ -67,23 +66,41 @@
+    {% endif %}
+  </div>
+
+  <h1{{ title_attributes.addClass('news-article__title') }}>
```
I left this item in in my PR, but we have to be careful about it as the page title is already printed in the page title block, so might appear twice in some pages. 

My preference is to leave it here and to _not_ use the page title block for node pages. That might mean updating the config to only show the page title block on some content types, or (more sustainably) to put a conditional around the page title region in `page.html.twig` to not load the page title if the page type is `node`.

danchamp replied:
I've been suppressing the page title block on the London Councils site to overcome this which isn't ideal.

It's a bit bound up in the theme page structure, where the page title block and therefore the page heading currently sits outside of any meaningful container. Discussion here: https://github.com/localgovdrupal/localgov_theme/issues/149

markconroy replied:
It's a Drupal hangover. I really don't like it. I'd much prefer if we had the page titles printed in the node templates.

In my theme in work (not for LocalGov), I have this in the page.html.twig file:

```
{% if not node %}
  {{ page.title }}
{% endif %}
```

And that ensures that the page block is not printed on any node pages. So there is no config overrides needed per content type. That's an approach we could pursue here. Or perhaps add it as a theme setting on the theme's setting pages, to make sure we don't break any current sites.

markconroy replied:
That approach means we need to maintain that code and keep it updated if
- **we** add any new content types, and/or
- **any council** using localgovdrupal adds any new content types (which will mean custom modules to override custom modules)

I'd probably prefer to have it set that:
- Page Title block does not display on any node page
- We have a `node--full.html.twig` as the default for all nodes that don't have a more specific template (such as `node--news-article--full.html.twig`) and print the page title in that template

That is a more scalable approach I think, rather than something that needs maintenance and re-thinking about each time new things are added.

danchamp replied:
Understood, that makes a lot of sense for future maintenance.

Is there any way we can achieve it in this module within the existing bounds of the core module and theme?

markconroy replied:
You idea to follow the approach of the subsites module would work, but then it would mean we have _some_ code hiding the page title block in the news module and _some_ code hiding the page title block in the subsites module. If events or some other module wants to do the same thing, you can see how it will start getting problematic at scale to have the same function spread across multiple modules.

My suggestion in the short term might be to add news as one of the content types in the list in the subsites module, then create an issue for us to tackle this is a sustainable manner - in localgov_core or localgov_skeleton, or somewhere else. I don't think this is a hard issue to solve, but it might take some coordinating with the other councils if it's removing a block that they are expecting.

danchamp replied:
I agree with the approach, but I'm not keen on the suggested short term solution - we'd have to make the subsites module as a dependency for the news module if we add the news content types there, which feels messier than using the mechanism that localgov_core provides for this purpose.

I've added an event subscriber to news for now: https://github.com/localgovdrupal/localgov_news/pull/15/commits/4bc769e4ca9e49ca9338c16569f51ccc2defa446

This lets us get closer to a release for news and can be removed later.

markconroy reviewed (commented):
This is looking really good now.

danchamp committed:
Use translatable string for categories ARIA label

danchamp committed:
Set news article date format in view display

danchamp committed:
Remove CSS attachment

danchamp committed:
Update view display config

danchamp committed:
Add missing class to news category list items

danchamp committed:
Merge branch 'master' into fix/13-category-display

danchamp committed:
Hide page header block on news articles

danchamp committed:
Update news article datetime attribute to use a valid format

See https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-time-element

danchamp committed:
Fix test name

danchamp committed:
Add event subscriber service

danchamp committed:
Fix coding standards

danchamp committed:
Update test.yml from master

danchamp committed:
Refactor templates

Add / update base styling
Suppress 'Show more' button on first load if no more articles to display

andybroomfield commented:
@danchamp Is this finished / Reviewable or is it being worked on still?

andybroomfield commented:
I'm just seeing blank tags instead of categories when I check a news article.
<img width="1197" alt="Screenshot 2021-05-15 at 4 24 34 pm" src="https://user-images.githubusercontent.com/1467480/118368487-22e5ab00-b59a-11eb-829e-2a4f5a8b2e67.png">

andybroomfield reviewed (changes requested):
Managed to resolve above, but required reinstall of the module, so this is going to require a db update to catch it

markconroy commented:
@andybroomfield looking at the classes in your inspector screenshot, it looks like all the code has not been pushed just yet.

We're working on news at the moment, so this should be updated this afternoon.

danchamp committed:
Refactor:

Replace infinite scroll with view block
Remove default image field
Replace news image field with field_media_image
Remove summary pseudo-field
Update tests
Update readme

danchamp commented:
Superceded by #31
